### ISSUE TITLE
Future of the library

### ISSUE THREAD

**Issue by tailhook** [tailhook/rust-argparse#51]
I was used to thinking that this library is not needed anymore because clap and structopt do the job just well. But there are valid [points](https://github.com/tailhook/rust-argparse/pull/50#issuecomment-455808876) by @d-e-s-o, the most important of it is that clap is very large in size.

On the other hand, I very much like the *model* structopt provides (i.e. parse in the structure instead of complex borrows). The actual *semantics* of structopt annotations are sometimes ugly though because it is a thin wrapper around clap (with different actual syntax).

So we have a few options:
1. Continue support of argparse as is (we need to write rust-style docs, as well as fix few current bugs, requiring breaking changes)
2. Make structopt-like wrapper (I mean derive) on top of argparse-based parser
3. Try to optimize clap+structopt if possible

**Comment by d-e-s-o** [comment]
I have updated numbers now that include `structopt`. Note that total numbers are smaller than those mentioned here https://github.com/tailhook/rust-argparse/pull/50 because I updated to Rust 1.32 which changes the allocator used (removing roughly 324 KiB from binaries according to measurements I've done in the past).

(unit is bytes)

With default settings:
structopt: 862432
clap: 866528
getopts: 358624

When optimizing for size (opt-level = "z", lto = true, codegen-units = 1, incremental = false):
structopt: 575712 (https://github.com/d-e-s-o/dictcc-cli/commit/6cad765c1f18267249eefeb09b158fcf1c4fb458)
clap: 579808 (https://github.com/d-e-s-o/dictcc-cli/commit/0a48fe9579139970d1faea6a941a3da21e846007)
getopts: 280800 (https://github.com/d-e-s-o/dictcc-cli/commit/06c7e738fa3637df426fd432d57529a4a2189f4d)

We can see that while `structopt` does not add anything over `clap` (in fact, it reduces the overall size slightly), it still plays in the same ball park. For small programs, I would argue, the increases that `clap` and `structopt` incur are quite significant.

**Comment by tailhook** [comment]
It's still not clear how much argparse adds. In my expericence, it's ~100Kb, which isn't too small either :(

Other than that, I haven't had time to investigate it myself, yet...

**Comment by d-e-s-o** [comment]
I read through the two articles that the author of `clap` has published. I doubt we can do much on this front (for the following I equate `clap` and `structopt` as the size of the former directly influences the size of the latter by virtue of `structopt` just being a wrapper; we have seen very similar binary size characteristics).
The size truly seems to come down to code (`.text`) and the author already has made effort to get this down. It seems unlikely that someone less familiar with the code base can change that much for the better, given that there was already work on this front that picked the low hanging fruits.

**Comment by cronokirby** [comment]
I like this crate more than `structopt`, mainly because it's quite a bit more lightweight, and straight to the point, since it avoids dabbling in macros.

**Comment by ijackson** [comment]
I am using this library primarily because `clap` (and therefore `structopt`) does not understand (and so cannot be used to implemnet) conventional Unix command line parsing.  For example, it is usual for a Unix program to understand later options as overriding earlier ones, but clap treats them as conflicting.  You can't even do something as simple as have a boolean `--thing` `--no-thing` pair where `--no-thing` can be used to override an earlier `--thing`.  There are other anomalies and `clap` in particular lacks the flexibility that `argparse` provices.

I would love to have something more declarative, structopt-style.  I don't care about the code size very much :-).

To go back to this library, my view is:
 * The documentation needs a lot of work.
 * Some of the liftetime requirements are pretty annoying.  For example, `add_option` et al insists on taking `&' parser str` for help text etc.  This makes it more awkward than necessary to have programatically-generated strings.  IMO these parameters should call be `Into<String>` and argparse should make a copy.  The extra heap allocations are trivial in cost compared to program execution so we should provide an easier interface and not try to hyperoptimise by taking references to avoid copies.
 * Different options affecting the same setting (`Ref`) should be able to have different `metavar`s.
 * There are some aspects of the API a that are un-idiomatic (at least by modern standards).  For example, `StoreOption` ought probably to be called `StoreSome`.

Some of these changes would be API breaks.

**Comment by tailhook** [comment]
@ijackson I agree with most of what you have said. I think this was my first project in Rust and dates back to Rust ~0.7, so certainly needs updating. And I'm also not happy with some of the clap's limitations. However, I do not have enough time to work on this now.

**Comment by ijackson** [comment]
> @ijackson I agree with most of what you have said. I think this was my first project in Rust and dates back to Rust ~0.7, so certainly needs updating. And I'm also not happy with some of the clap's limitations,

Well I think you have done well for your first project then!  Thank you :-).

> However, I do not have enough time to work on this now.

I may find time to do so.  If I do, should I submit MRs, or pick a new name and fork it, or what?  I guess the first step would be an MR with a branch containing proposed changes.

**Comment by sunshowers** [comment]
> You can't even do something as simple as have a boolean `--thing` `--no-thing` pair where `--no-thing` can be used to override an earlier `--thing`. 

BTW this can be done with [overrides_with](https://docs.rs/clap/latest/clap/struct.Arg.html#method.overrides_with) in clap.

**Comment by sztomi** [comment]
I know this is not exactly a fresh thread, but I just did an experiment and wanted to share. A simple command-line with just four positional arguments:

```rs
use std::path::PathBuf;

use argparse::{ArgumentParser, Store};


#[derive(Default, Debug)]
struct CommandLine {
    sacred_ref: String,
    origin_repo_path: PathBuf,
    replica_repo_path: PathBuf,
    root_subdir: PathBuf,
}

fn main() {
    let mut cli = CommandLine::default();
    {
        let mut parser = ArgumentParser::new();
        parser.refer(&mut cli.origin_repo_path)
            .add_argument("origin_repo_path", Store, "Path to the origin repo")
            .required();
        parser.refer(&mut cli.replica_repo_path)
            .add_argument("replica_repo_path", Store, "Path to the replica repo")
            .required();
        parser.refer(&mut cli.root_subdir)
            .add_argument(
                "root_subdir",
                Store,
                "Relative path inside origin that will be the root of replica")
            .required();
        parser.refer(&mut cli.sacred_ref)
            .add_argument(
                "sacred_ref",
                Store,
                "Sacred ref to start replication from in the history")
            .required();
        parser.parse_args_or_exit();
    }

    println!("{:?}", &cli);
}
```

```
cargo b --release  1.36s user 0.20s system 122% cpu 1.274 total
```

In release, this yields a 491KB binary on macOS arm64. 

A reasonably similar CLI using clap:

```rs
use std::path::PathBuf;

use clap::Parser;

#[derive(Parser, Debug)]
#[command(name = "subgit")]
#[command(about = "A tool for replicating subdirectories from git repositories")]
struct CommandLine {
    #[arg(help = "Path to the origin repo")]
    origin_repo_path: PathBuf,
    
    #[arg(help = "Path to the replica repo")]
    replica_repo_path: PathBuf,
    
    #[arg(help = "Relative path inside origin that will be the root of replica")]
    root_subdir: PathBuf,
    
    #[arg(help = "Sacred ref to start replication from in the history")]
    sacred_ref: String,
}

fn main() {
    let cli = CommandLine::parse();
    
    println!("{:?}", &cli);
}
```

```
cargo b --release  16.56s user 0.98s system 354% cpu 4.955 total
```

In release, this yields a 917KB binary on macOS arm64. 

I think `argparse` has its place in the Rust ecosystem, and it would be a shame to fully retire it. Coming from Python, I like its approach and I like the fact that there is much less magic happening (which is also probably a big contributor to the vastly superior compilation time). Now, I'm sure there are applications with extremely complex command lines where clap scales better, but for me, `argparse` easily wins.
